substrate/frame/support/src/migrations.rs

```diff
@@ -345,3 +348,573 @@ impl<P: Get<&'static str>, DbWeight: Get<RuntimeDbWeight>> frame_support::traits
+	///
+	/// This should not happen unless (1) the migration itself returned an invalid cursor in a
+	/// previous iteration, (2) the storage got corrupted or (3) there is a bug in the caller's
+	/// code.
+	InvalidCursor,
+	/// The migration encountered a permanent error and cannot continue.
+	Failed,
+}
+
+/// Notification handler for status updates regarding Multi-Block-Migrations.
+#[impl_trait_for_tuples::impl_for_tuples(8)]
+pub trait MigrationStatusHandler {
+	/// Notifies of the start of a runtime migration.
+	fn started() {}
+
+	/// Notifies of the completion of a runtime migration.
+	fn completed() {}
+}
+
+/// Handles a failed runtime migration.
+///
+/// This should never happen, but is here for completeness.
+pub trait FailedMigrationHandler {
+	/// Infallibly handle a failed runtime migration.
+	///
+	/// Gets passed in the optional index of the migration in the batch that caused the failure.
+	/// Returning `None` means that no automatic handling should take place and the callee decides
+	/// in the implementation what to do.
+	fn failed(migration: Option<u32>) -> Option<FailedUpgradeHandling>;
+}
+
+impl FailedMigrationHandler for () {
+	fn failed(_migration: Option<u32>) -> Option<FailedUpgradeHandling> {
+		Some(FailedUpgradeHandling::KeepStuck)
+	}
+}
+
+/// How to proceed after a runtime upgrade failed.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum FailedUpgradeHandling {
+	/// Resume extrinsic processing of the chain. This will not resume the upgrade.
+	///
+	/// This should be supplemented with additional measures to ensure that the broken chain state
+	/// does not get further messed up by user extrinsics.
+	ForceUnstuck,
+	/// Set the cursor to `Stuck` and keep blocking extrinsics.
+	KeepStuck,
+}
+
+/// Something that can do multi step migrations.
+pub trait MultiStepMigrator {
+	/// Hint for whether [`Self::step`] should be called.
+	fn ongoing() -> bool;
+
+	/// Do the next step in the MBM process.
+	///
+	/// Must gracefully handle the case that it is currently not upgrading.
+	fn step() -> Weight;
+}
+
+impl MultiStepMigrator for () {
+	fn ongoing() -> bool {
+		false
+	}
+
+	fn step() -> Weight {
+		Weight::zero()
+	}
+}
+
+/// Multiple [`SteppedMigration`].
+pub trait SteppedMigrations {
+	/// The number of migrations that `Self` aggregates.
+	fn len() -> u32;
+
+	/// The `n`th [`SteppedMigration::id`].
+	///
+	/// Is guaranteed to return `Some` if `n < Self::len()`.
+	fn nth_id(n: u32) -> Option<Vec<u8>>;
+
+	/// The [`SteppedMigration::max_steps`] of the `n`th migration.
+	///
+	/// Is guaranteed to return `Some` if `n < Self::len()`.
+	fn nth_max_steps(n: u32) -> Option<Option<u32>>;
+
+	/// Do a [`SteppedMigration::step`] on the `n`th migration.
+	///
+	/// Is guaranteed to return `Some` if `n < Self::len()`.
+	fn nth_step(
+		n: u32,
+		cursor: Option<Vec<u8>>,
+		meter: &mut WeightMeter,
+	) -> Option<Result<Option<Vec<u8>>, SteppedMigrationError>>;
```

Comment:
some feedback on our own experience building mbm:

We opted for not taking a meter but having a `fn max_step_weight` that returns the maximum amount of weight a single step can take. This moves the burden of checking whether or not we have enough weight to perform any work to the framework rather than the migration itself

https://github.com/paritytech/polkadot-sdk//blob/6fe7ac4cc8a9004c14873fc477de81b83a1b17dd/substrate/frame/contracts/src/migration.rs#L110-L116

Reply:
>We opted for not taking a meter but having a fn max_step_weight that returns the maximum amount of weight a single step can take

I am not sure about this since it makes it less flexible. The migration may decide to make larger steps when there is more weight available to be more efficient.  
Yes the code in the migration has not to do this task, but it also has more freedom of how to use the weight...

Reply:
But it probably is a good idea to add a `min_weight` to the interface to ensure that it at least can always make *some* progress.

Reply:
> The migration may decide to make larger steps when there is more weight available to be more efficient

You will need to check the weight regardless no? So why not moving this responsibility to the framework?

In our case the migration framework verify there is enough weight before running a new step and do that in a loop until all the weight is exhausted